verilog/tools/ls/verilog-language-server_test.cc

```diff
@@ -0,0 +1,485 @@
+
+// TODO (glatosinski) for JSON messages use types defined in lsp-protocol.h
+
+using namespace nlohmann;
+
+class VerilogLanguageServerTest : public ::testing::Test {
+ public:
+  // Sends initialize request from client mock to the Language Server.
+  // It does not parse the response nor fetch it in any way (for other
+  // tests to check e.g. server/client capabilities).
+  absl::Status InitializeCommunication() {
+    const absl::string_view initialize =
+        R"({ "jsonrpc": "2.0", "id": 1, "method": "initialize", "params": null })";
+
+    SetRequest(initialize);
+    return ServerStep();
+  }
+
+  // Runs SetRequest and ServerStep, returning the status from the
+  // Language Server
+  absl::Status SendRequest(absl::string_view request) {
+    SetRequest(request);
+    return ServerStep();
+  }
+
+  // Returns the latest responses from the Language Server
+  std::string GetResponse() {
+    std::string response = response_stream_.str();
+    response_stream_.str("");
+    response_stream_.clear();
+    return response;
+  }
+
+  // Returns response to textDocument/initialize request
+  std::string GetInitializeResponse() { return initialize_response_; }
+
+ private:
+  // Wraps request for the Language Server in RPC header
+  void SetRequest(absl::string_view request) {
+    request_stream_.clear();
+    request_stream_.str(
+        absl::StrCat("Content-Length: ", request.size(), "\r\n\r\n", request));
+  }
+
+  // Performs single VerilogLanguageServer step, fetching latest request
+  absl::Status ServerStep() {
+    return server_->Step([this](char *message, int size) -> int {
+      request_stream_.read(message, size);
+      return request_stream_.gcount();
+    });
+  }
+
+  // Sets up the testing environment - creates Language Server object and
+  // sends textDocument/initialize request.
+  // It stores the response in initialize_response field for further processing
+  void SetUp() override {
+    server_ = std::make_unique<VerilogLanguageServer>(
+        [this](absl::string_view response) { response_stream_ << response; });
+
+    absl::Status status = InitializeCommunication();
+    EXPECT_TRUE(status.ok()) << "Failed to read request:  " << status;
+    initialize_response_ = GetResponse();
+  }
+
+  // Currently tested instance of VerilogLanguageServer
+  std::unique_ptr<VerilogLanguageServer> server_;
+
+  // Response from textDocument/initialize request - left for checking e.g
+  // server capabilities
+  std::string initialize_response_;
+
+  // Stream for passing requests to the Language Server
+  std::stringstream request_stream_;
+
+  // Stream for receiving responses from the Language Server
+  std::stringstream response_stream_;
+};
+
+// Verifies textDocument/initialize request handling
+TEST_F(VerilogLanguageServerTest, InitializeRequest) {
+  std::string response_str = GetInitializeResponse();
+  json response = json::parse(response_str);
+
+  ASSERT_EQ(response["id"], 1) << "Response message ID invalid";
```

Comment:
Typically, we use `ASSERT_xx()` in tests where we want to stop the test and don't execute anything afterwards (e.g. because we know now everything following will just crash), and `EXPECT_xx()` to verify that a particular condition is met, but it will keep going. That way we can see multiple things going on.

https://testing.googleblog.com/2008/07/tott-expect-vs-assert.html

So many `ASSERT_xx()` in this test can be changed to `EXPECT_xx()`. Probably only the `SendRequest()` need to be asserted (as anything afterwards does not make sense if that failed), but all the comparison if particular values are set as they are can be EXPECT.

Something for a TODO later.